File: hw/paillacho/U3/HW22Observer/ObserverPython/src/model/stock.py
"""
Subject (Observable) in the Observer pattern
@author Gerald Astudillo
"""

from abc import ABC, abstractmethod

class Stock(ABC):
    def __init__(self):
        self._symbol = None
        self._price = None
        self._investors = []

    def add_observer(self, investor):
        """Add an observer to the list"""
        self._investors.append(investor)

    def delete_observer(self, investor):
        """Remove an observer from the list"""
        if investor in self._investors:
            self._investors.remove(investor)

    def notify_observers(self, args):
        """Notify all observers about changes"""
        for investor in self._investors:
            investor.update(self, args)

    @abstractmethod
    def get_symbol(self):
        """Abstract method to get symbol"""
        pass

    @abstractmethod
    def get_price(self):
        """Abstract method to get price"""
        pass

File: hw/astudillo/U3/HW22Observer/Observer/src/main/java/ec/edu/espe/observer/model/Stock.java
package ec.edu.espe.observer.model;

import java.util.ArrayList;
import java.util.Iterator;

/**
 * Subject (Observable) in the Observer pattern
 * @author Gerald Astudillo
 */
public abstract class Stock {
    protected String symbol;
    protected double price;
    private ArrayList<IInvestor> investors = new ArrayList<>();

    public Stock() {}

    public void addObserver(IInvestor investor) {
        investors.add(investor);
    }

    public void deleteObserver(IInvestor investor) {
        investors.remove(investor);
    }

    public void notifyObservers(Object args) {
        Iterator<IInvestor> i = investors.iterator();
        while (i.hasNext()) {
            IInvestor investor = i.next();
            investor.update(this, args);
        }
    }

    public abstract String getSymbol();
    public abstract double getPrice();
}
